### ISSUE TITLE
loss.backward() cause cpu memory leak

### ISSUE THREAD

**Issue by nido4010** [pytorch/pytorch#1217]
Hello, thank you for pytorch.
I am studying tutorial
When I run cifar10_tutorial.py code using gpu, I found the memory leak

When I delete loss.backward(), memory leak doesn't occur

please, resolve this problem

**Comment by nido4010** [comment]
Installation enviroment
Install method : binary pip
Python :2.7
Cuda: 8.0
Cudnn : 5.1
Os : linux mint 18.1

**Comment by soumith** [comment]
I'm not sure what you mean by "memory leak", can you elaborate?

**Comment by nido4010** [comment]
model code
`
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(3, 128, 5)
        self.pool  = nn.MaxPool2d(2, 2)
        self.conv2 = nn.Conv2d(128, 128, 5)
        self.fc1   = nn.Linear(128 * 5 * 5, 120)
        self.fc2   = nn.Linear(120, 84)
        self.fc3   = nn.Linear(84, 10)

    def forward(self, x):
        x = self.pool(F.relu(self.conv1(x)))
        x = self.pool(F.relu(self.conv2(x)))
        x = x.view(-1, 128 * 5 * 5)
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x

net = Net().cuda()`

loss and optimizer code
`
criterion = nn.CrossEntropyLoss().cuda()
optimizer = optim.SGD(net.parameters(), lr=0.001, momentum=0.9)`

train code

`for epoch in range(5):  # loop over the dataset multiple times

    running_loss = 0.0
    for i, data in enumerate(trainloader, 0):
        # get the inputs
        inputs, labels = data

        # wrap them in Variable
        inputs, labels = Variable(inputs.cuda()), Variable(labels.cuda())

        # zero the parameter gradients
        optimizer.zero_grad()

        # forward + backward + optimize
        outputs = net(inputs)

        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()

        # print statistics
        running_loss += loss.data[0]
        if i % 2000 == 1999:    # print every 2000 mini-batches
            print('[%d, %5d] loss: %.3f' % (epoch+1, i+1, running_loss / 2000))
            running_loss = 0.0

    os.system('free')
print('Finished Training')
`
I check  cpu memory every epoch. I found memory leak(about 70MB) every epoch like below

>               total        used        free      shared  buff/cache   available
Mem:        8111888     3346952     1136564       85236     3628372     4363272
Swap:       3998716        4616     3994100
              total        used        free      shared  buff/cache   available
Mem:        8111888     3425516     1050540       92256     3635832     4277412
Swap:       3998716        4616     3994100
              total        used        free      shared  buff/cache   available
Mem:        8111888     3499164      976944       92012     3635780     4203940
Swap:       3998716        4616     3994100
              total        used        free      shared  buff/cache   available
Mem:        8111888     3570928      911648       85596     3629312     4138584
Swap:       3998716        4616     3994100
              total        used        free      shared  buff/cache   available
Mem:        8111888     3644624      837764       85716     3629500     4064816
Swap:       3998716        4616     3994100

**Comment by apaszke** [comment]
Your comment doesn't include all the necessary parts that would allow us to run the script. Also, don't paste large chunks of code along with screenshots in comments, it only makes the thread harder to follow. If you suspect you're seeing a mem leak, please upload the full script to gist.github.com and past a link in here.

**Comment by nido4010** [comment]
<script src="https://gist.github.com/nido4010/20e9fb05cbaecf662ce34a1d31ef46d4.js"></script>

**Comment by nido4010** [comment]
sorry, I am a newbie here

Here is the code
https://gist.github.com/nido4010/20e9fb05cbaecf662ce34a1d31ef46d4

**Comment by soumith** [comment]
I ran the tutorial (I disabled plotting because i didnt have a display):

```
python cifar10_tutorial.py
Files already downloaded and verified
Files already downloaded and verified
horse   cat   cat truck
epoch 1
              total        used        free      shared  buff/cache   available
Mem:       32720992     2660788    23783680       31308     6276524    29459632
Swap:      33460220           0    33460220
epoch 2
              total        used        free      shared  buff/cache   available
Mem:       32720992     2661224    23782668       31344     6277100    29458728
Swap:      33460220           0    33460220
epoch 3
              total        used        free      shared  buff/cache   available
Mem:       32720992     2661112    23782528       31272     6277352    29458632
Swap:      33460220           0    33460220
```

This is my output, memory seems rock-stable.
There are instances by the memory allocator used on the system where it keeps blocks around and doesn't free them. Maybe your default memory allocator keeps memory around?
Either that, or maybe plotting is adding unfreeable memory? Can you check by commenting out plotting? Comment out the line `plt.imshow(np.transpose(npimg, (1, 2, 0)))`

**Comment by nido4010** [comment]
Thank you for your reply.
Unfortunately, memory leak occur not only in this tutorial code but also in my convolution neural network code. I can't find the reason. 
So, I will reinstall osn and I will try to run this code again

**Comment by nido4010** [comment]
I reinstall OS and comment out the line plt.imshow(np.transpose(npimg, (1, 2, 0)))
but memory leak occur
`epoch 1
              total        used        free      shared  buff/cache   available
Mem:        8111900     2603728     3408092       66580     2100080     5134764
Swap:      15998972           0    15998972
epoch 2
              total        used        free      shared  buff/cache   available
Mem:        8111900     2670360     3341752       66272     2099788     5068532
Swap:      15998972           0    15998972
epoch 3
              total        used        free      shared  buff/cache   available
Mem:        8111900     2741104     3270544       66648     2100252     4997360
Swap:      15998972           0    15998972
epoch 4
              total        used        free      shared  buff/cache   available
Mem:        8111900     2811956     3199772       66512     2100172     4926620
Swap:      15998972           0    15998972
epoch 5
              total        used        free      shared  buff/cache   available
Mem:        8111900     2886968     3124680       66576     2100252     4851568
Swap:      15998972           0    15998972
Finished Training`

I use linux mint 18.1 OS and install pytorch using binary pip.
What OS did you use? and what method did you use for installation?

**Comment by nido4010** [comment]
I found out the cause.
It is the problem related to cudnn.
I reopen this issue to explain the reason
